plugins/autocomplete/plugin.js

```diff
@@ -857,94 +857,165 @@
+				// |                  +--------------+           |
+				// | caret position - █                          |
+				// |                                             |
+				// =============================================== - bottom window border
+				// |                                             |
+				// |                                             |
+				// +---------------------------------------------+
+				var viewExceedsBottomViewport = ( caretRect.bottom + viewHeight ) > ( windowHeight + verticalScroll );
+
+				if ( !( viewHeight > spaceBelow && viewHeight < spaceAbove ) && viewExceedsBottomViewport ) {
+					return caretRect.top - viewHeight;
+				}
+
+				// As a default, keep the view inside the editor viewport.
+				// +---------------------------------------------+
+				// |       editor viewport                       |
+				// |                                             |
+				// |                                             |
+				// |                                             |
+				// |     █ - caret position                      |
+				// |     +--------------+                        |
+				// |     |     view     |                        |
+				// |     +--------------+                        |
+				// |                                             |
+				// |                                             |
+				// +---------------------------------------------+
+				return Math.min( editorViewportRect.bottom, caretRect.bottom );
 			}
 
-			// If the caret position is above the view - keep it at the top edge.
-			// +---------------------------------------------+
-			// |																						 |
-			// |     █ - caret position                      |
-			// |                                             |
-			// +-----+==============+------------------------+
-			// |     |              |                        |
-			// |     |     view     |                        |
-			// |     |              |                        |
-			// |     +--------------+                        |
-			// |																						 |
-			// |       editor viewport                       |
-			// +---------------------------------------------+
-			if ( editorViewportRect.top > rect.top ) {
-				top = Math.max( rect.bottom, editorViewportRect.top );
+			// Bounding rect where the view should fit (visible editor viewport).
+			function getEditorViewportRect( editor ) {
+				var editable = editor.editable();
+
+				// iOS classic editor has different viewport element (#1910).
+				if ( CKEDITOR.env.iOS && !editable.isInline() ) {
+					return iOSViewportElement( editor ).getClientRect( true );
+				} else {
+					return editable.isInline() ? editable.getClientRect( true ) : editor.window.getFrame().getClientRect( true );
+				}
 			}
 
-			this.element.setStyles( {
-				left: rect.left + 'px',
-				top: top + 'px'
-			} );
+			function setHorizontalPosition( options ) {
+				var leftPosition = options.leftPosition,
```

Comment:
```suggestion
              var caretLeftPosition = options.leftPosition,
```

?

Reply:
I didn't add `caret` prefix here because according to docs it doesn't have to be a caret:
```
rect : Object
    Represents the position of a vertical (e.g. a caret) line relative to which the panel should be positioned.
```
but I see I forgot about it in the `setVerticalPosition()` function anyway 🙈 And since it is rather hard to imagine something else, I agree we should go with `caretLeftPosition` here.